SIGSEGV: invalid memory address or nil pointer dereference in DeviceNames

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.16.0 and lower than v1.17.0

### What happened?

Since upgrading to Cilium 1.16.0, I sporadically observe Cilium Agent container crash/restarts caused by below nil pointer error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x90 pc=0x2642570]

goroutine 984 [running]:
github.com/cilium/cilium/pkg/datapath/types.(*LocalNodeConfiguration).DeviceNames(...)
	/go/src/github.com/cilium/cilium/pkg/datapath/types/node.go:165
github.com/cilium/cilium/pkg/datapath/linux/config.(*HeaderfileWriter).WriteEndpointConfig(0x4002798000, {0x3dce300, 0x4000da29e0?}, 0x0, {0x3e531f0, 0x4000468c08})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/config/config.go:1045 +0xf0
github.com/cilium/cilium/pkg/datapath/loader.(*loader).WriteEndpointConfig(0x4000468c08?, {0x3dce300?, 0x4000da29e0?}, {0x3e531f0?, 0x4000468c08?})
	/go/src/github.com/cilium/cilium/pkg/datapath/loader/loader.go:698 +0x50
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile(0x4000468c08, {0x40029fbac0, 0x1a})
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:189 +0x540
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).syncEndpointHeaderFile(0x4000468c08, {0x4001249720, 0x1, 0x1})
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2450 +0x294
github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter(0x4002532960)
	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:201 +0x464
created by github.com/cilium/cilium/pkg/trigger.NewTrigger in goroutine 1
	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:122 +0x1b4
```

### How can we reproduce the issue?

N/A

### Cilium Version

1.16.0

### Kernel Version

6.1.94

Bottlerocket OS 1.20.4 (aws-k8s-1.29)
containerd://1.6.31

### Kubernetes Version

v1.29.4-eks-036c24b

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell
time="2024-07-25T18:20:29Z" level=info msg="Subscribed to CiliumNode custom resource" name=ip-10-176-21-6.eu-central-1.compute.internal subsys=ipam
time="2024-07-25T18:20:29Z" level=info msg="Creating or updating CiliumNode resource" node=ip-10-176-21-6.eu-central-1.compute.internal subsys=nodediscovery
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x90 pc=0x2642570]

goroutine 984 [running]:
github.com/cilium/cilium/pkg/datapath/types.(*LocalNodeConfiguration).DeviceNames(...)
	/go/src/github.com/cilium/cilium/pkg/datapath/types/node.go:165
github.com/cilium/cilium/pkg/datapath/linux/config.(*HeaderfileWriter).WriteEndpointConfig(0x4002798000, {0x3dce300, 0x4000da29e0?}, 0x0, {0x3e531f0, 0x4000468c08})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/config/config.go:1045 +0xf0
github.com/cilium/cilium/pkg/datapath/loader.(*loader).WriteEndpointConfig(0x4000468c08?, {0x3dce300?, 0x4000da29e0?}, {0x3e531f0?, 0x4000468c08?})
	/go/src/github.com/cilium/cilium/pkg/datapath/loader/loader.go:698 +0x50
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile(0x4000468c08, {0x40029fbac0, 0x1a})
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:189 +0x540
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).syncEndpointHeaderFile(0x4000468c08, {0x4001249720, 0x1, 0x1})
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2450 +0x294
github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter(0x4002532960)
	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:201 +0x464
created by github.com/cilium/cilium/pkg/trigger.NewTrigger in goroutine 1
	/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:122 +0x1b4
```


### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -649,6 +649,11 @@ func (e *Endpoint) regenerateBPF(regenContext *regenerationContext) (revnum uint
 		return 0, fmt.Errorf("unable to regenerate policy because PolicyMap synchronization failed: %w", err)
 	}
 
+	// Initialize (if not done yet) the DNS history trigger to allow DNS proxy to trigger
+	// updates to endpoint headers. The initialization happens here as at this point
+	// datapath is ready to process the trigger.
+	e.initDNSHistoryTrigger()
+
 	return datapathRegenCtxt.epInfoCache.revision, err
 }
 

--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -273,7 +273,7 @@ type Endpoint struct {
 
 	// dnsHistoryTrigger is the trigger to write down the ep_config.h to make
 	// sure that restores when DNS policy is in there are correct
-	dnsHistoryTrigger *trigger.Trigger
+	dnsHistoryTrigger atomic.Pointer[trigger.Trigger]
 
 	// state is the state the endpoint is in. See setState()
 	state State
@@ -587,8 +587,6 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
 		properties:       map[string]interface{}{},
 	}
 
-	ep.initDNSHistoryTrigger()
-
 	ctx, cancel := context.WithCancel(context.Background())
 	ep.aliveCancel = cancel
 	ep.aliveCtx = ctx
@@ -601,16 +599,22 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
 }
 
 func (e *Endpoint) initDNSHistoryTrigger() {
+	if e.dnsHistoryTrigger.Load() != nil {
+		// Already initialized, bail out.
+		return
+	}
+
 	// Note: This can only fail if the trigger func is nil.
-	var err error
-	e.dnsHistoryTrigger, err = trigger.NewTrigger(trigger.Parameters{
+	trigger, err := trigger.NewTrigger(trigger.Parameters{
 		Name:        "sync_endpoint_header_file",
 		MinInterval: 5 * time.Second,
 		TriggerFunc: e.syncEndpointHeaderFile,
 	})
 	if err != nil {
 		log.WithField(logfields.EndpointID, e.ID).WithError(err).Error("Failed to create the endpoint header file sync trigger")
+		return
 	}
+	e.dnsHistoryTrigger.Store(trigger)
 }
 
 // CreateIngressEndpoint creates the endpoint corresponding to Cilium Ingress.
@@ -1239,8 +1243,8 @@ func (e *Endpoint) leaveLocked(proxyWaitGroup *completion.WaitGroup, conf Delete
 	e.controllers.RemoveAll()
 	e.cleanPolicySignals()
 
-	if e.dnsHistoryTrigger != nil {
-		e.dnsHistoryTrigger.Shutdown()
+	if trigger := e.dnsHistoryTrigger.Swap(nil); trigger != nil {
+		trigger.Shutdown()
 	}
 
 	if e.ConntrackLocalLocked() {
@@ -2448,8 +2452,8 @@ func (e *Endpoint) syncEndpointHeaderFile(reasons []string) {
 // SyncEndpointHeaderFile triggers the header file sync to the ep_config.h
 // file. This includes updating the current DNS History information.
 func (e *Endpoint) SyncEndpointHeaderFile() {
-	if e.dnsHistoryTrigger != nil {
-		e.dnsHistoryTrigger.Trigger()
+	if trigger := e.dnsHistoryTrigger.Load(); trigger != nil {
+		trigger.Trigger()
 	}
 }
 

